### PR TITLE
[ROCm][Perf] Sparse-attention Triton kernel optimizations for MiniMax-M3

### PR DESCRIPTION
**Pull request by yueliu14**
## Summary

Two independent, numerically-equivalent Triton-kernel optimizations for MiniMax-M3 sparse attention on AMD MI300X (CDNA3 / gfx942):

1. **`_gqa_sparse_fwd_kernel`** (`sparse_attn.py`) — prefill block-sparse GQA. Splits the per-block KV loop into `SUB_K = BLOCK_SIZE_K / 4` sub-tiles and launches with tuned MFMA parameters (`num_warps=1, matrix_instr_nonkdim=16, kpack=2`), right-sizing the QK/PV tiles and raising occupancy.

2. **`_decode_index_score_kernel`** (`index_topk.py`) — decode indexer block scoring. Specializes the `BLOCK_SIZE_HQ == 1` decode path: the original `[N,D] × [D,1]` `tl.dot` is a degenerate GEMV that yields one useful column out of the ≥16-wide MFMA tile (>90% wasted) and inflates accumulator VGPRs; replaced with a vectorized fp32 multiply + reduce. The multi-head / spec-decode path (`BLOCK_SIZE_HQ > 1`) is unchanged and keeps the original `tl.dot`.

Both optimizations are **gated behind `_IS_ROCM_GFX942`** (a `tl.constexpr` resolved at import time via `vllm.platforms.rocm.on_gfx942()`), so NVIDIA and non-gfx942 ROCm platforms are completely unaffected. Both are pure compute rewrites (flash-softmax reassociation / fp32 reduction); the math is unchanged and outputs are numerically identical to the current kernels.

## End-to-end results (MI300X, TP=8)

```bash
# server
vllm serve MiniMax-M3-MXFP8 --port 8901 \
    --tensor-parallel-size 8 --block-size 128 \
    --no-enable-prefix-caching --language-model-only --max-model-len 9472 \
    --attention-backend TRITON_ATTN \
    --tool-call-parser minimax_m3 --reasoning-parser minimax_m3 --enable-auto-tool-choice

# client — one run per concurrency C in {1,2,4,8,16,32,64,128};
# num-prompts = C*10 (capped 320, floored 20)
vllm bench serve --model MiniMax-M3-MXFP8 --backend vllm \
    --dataset-name random --random-input-len <ISL> --random-output-len <OSL> \
    --random-range-ratio 0.8 --num-prompts <N> --max-concurrency <C> \
    --request-rate inf --ignore-eos
```

A/B is identical on both arms; **only the two kernels differ**.

### ISL=8192, OSL=1024

| conc | baseline out tok/s | opt out tok/s | baseline total tok/s | opt total tok/s | baseline TTFT (ms) | opt TTFT (ms) | baseline TPOT (ms) | opt TPOT (ms) |
|---|---|---|---|---|---|---|---|---|
| 1 | 86.8 | 92.1 (1.06×) | 773.8 | 820.4 (1.06×) | 450.2 | 357.7 (1.26×) | 10.7 | 10.5 (1.02×) |
| 2 | 166.4 | 174.8 (1.05×) | 1482.5 | 1557.4 (1.05×) | 469.8 | 374.0 (1.26×) | 11.2 | 10.8 (1.04×) |
| 4 | 285.6 | 300.9 (1.05×) | 2567.5 | 2705.0 (1.05×) | 479.0 | 384.5 (1.25×) | 13.2 | 12.5 (1.05×) |
| 8 | 424.7 | 457.1 (1.08×) | 3777.5 | 4065.9 (1.08×) | 485.2 | 388.4 (1.25×) | 18.0 | 16.6 (1.08×) |
| 16 | 591.7 | 663.3 (1.12×) | 5339.5 | 5986.4 (1.12×) | 502.5 | 398.7 (1.26×) | 25.8 | 22.9 (1.13×) |
| 32 | 788.8 | 909.2 (1.15×) | 7053.5 | 8130.0 (1.15×) | 523.4 | 414.2 (1.26×) | 39.3 | 34.0 (1.16×) |
| 64 | 867.4 | 1053.2 (1.21×) | 7756.9 | 9418.2 (1.21×) | 887.1 | 680.7 (1.30×) | 70.9 | 58.2 (1.22×) |
| 128 | 1098.3 | 1361.2 (1.24×) | 9821.5 | 12172.0 (1.24×) | 1446.1 | 1103.0 (1.31×) | 100.2 | 80.6 (1.24×) |

### ISL=1024, OSL=1024

| conc | baseline out tok/s | opt out tok/s | baseline total tok/s | opt total tok/s | baseline TTFT (ms) | opt TTFT (ms) | baseline TPOT (ms) | opt TPOT (ms) |
|---|---|---|---|---|---|---|---|---|
| 1 | 93.2 | 95.0 (1.02×) | 184.7 | 188.3 (1.02×) | 110.8 | 111.3 (1.00×) | 10.6 | 10.4 (1.02×) |
| 2 | 179.6 | 186.0 (1.04×) | 356.0 | 368.6 (1.04×) | 144.7 | 142.5 (1.02×) | 10.8 | 10.6 (1.02×) |
| 4 | 317.5 | 314.2 (0.99×) | 638.1 | 631.5 (0.99×) | 144.4 | 145.6 (0.99×) | 12.1 | 11.8 (1.02×) |
| 8 | 518.4 | 525.0 (1.01×) | 1033.0 | 1046.1 (1.01×) | 148.2 | 150.7 (0.98×) | 15.0 | 14.8 (1.01×) |
| 16 | 809.3 | 829.5 (1.03×) | 1627.2 | 1667.9 (1.03×) | 154.8 | 155.9 (0.99×) | 19.3 | 18.8 (1.03×) |
| 32 | 1177.0 | 1202.3 (1.02×) | 2350.2 | 2400.8 (1.02×) | 167.3 | 167.7 (1.00×) | 26.7 | 26.1 (1.02×) |
| 64 | 1451.8 | 1490.4 (1.03×) | 2898.9 | 2976.1 (1.03×) | 208.1 | 253.1 (0.82×) | 42.9 | 42.0 (1.02×) |
| 128 | 2169.2 | 2266.6 (1.04×) | 4331.5 | 4526.1 (1.04×) | 357.5 | 335.5 (1.07×) | 51.5 | 49.7 (1.04×) |

### Key takeaways

- **ISL=8192**: output throughput **+5–24%** (scales with concurrency), TTFT **1.25–1.31×** faster across all concurrencies, TPOT up to **1.24×** at conc=128.
- **ISL=1024**: throughput and latency within noise (~1–4%), **no regression**.

## Accuracy (gsm8k)

No regression — scores are within noise of the baseline:

| filter | baseline | opt |
|---|---|---|
| flexible-extract | 0.9575 ± 0.0056 | 0.9575 ± 0.0056 |
| strict-match | 0.9583 ± 0.0055 | 0.9583 ± 0.0055 |

## Test plan

- **Per-kernel correctness + speedup sweeps** (optimized vs stock vs fp32 reference) across the shape / batch ranges above; outputs numerically identical (cos-dist shown per point).
- **End-to-end serving** on MiniMax-M3-MXFP8 (MI300X, TP=8) with full concurrency sweep (1→128) across ISL=8192/OSL=1024 and ISL=1024/OSL=1024. Server stable, zero failed requests.
- **Accuracy** (gsm8k lm-eval) — no regression.
- Both optimizations gated behind `_IS_ROCM_GFX942` — NVIDIA and non-gfx942 platforms execute the original unmodified kernels.

## Notes

- `matrix_instr_nonkdim` / `kpack` are ROCm Triton launch knobs; other backends are unaffected (the kwargs are consumed by the ROCm Triton compiler).

### CODE REVIEW

File: vllm/models/minimax_m3/common/ops/index_topk.py [hunk 1]
```diff
@@ -373,7 +373,16 @@ def _decode_index_score_kernel(
             + off_k[:, None] * stride_ik_pos
             + off_d * stride_ik_d,
         )  # [N,D]
-        kq = tl.dot(k, q)  # [N,HQ]
+        if BLOCK_SIZE_HQ == 1:
```
Comment (hongxiayang) [inline vllm/models/minimax_m3/common/ops/index_topk.py]: this PR touches common path for both nvidia and amd. should at least gate the changes if this is specific to rocm

Comment (hongxiayang) [reply vllm/models/minimax_m3/common/ops/index_topk.py]: does the optimization apply to gfx950? will it cause regression?

Comment (yueliu14) [reply vllm/models/minimax_m3/common/ops/index_topk.py]: Thanks for reviewing! Added rocm and gfx942 gating and updated ISL/OSL=1k/1k results

File: vllm/models/minimax_m3/common/ops/sparse_attn.py [hunk 2]
```diff
@@ -463,6 +529,13 @@ def minimax_m3_sparse_attn(
     topk = topk_idx.shape[-1]
     gqa_group_size = num_heads // num_kv_heads
     use_fp8 = kv_cache.dtype in _FP8_DTYPES
+    gfx942_kwargs: dict = {}
```
Comment (tjtanaa) [inline vllm/models/minimax_m3/common/ops/sparse_attn.py]: @yueliu14 

Change 1: Following this RFC https://github.com/vllm-project/vllm/issues/42770 , your changes is ROCm specific, so let's create a clone of these two op files into `https://github.com/vllm-project/vllm/tree/main/vllm/models/minimax_m3/amd/ops` 
After you clone into the `amd/ops` path , please update the import path for ROCm/AMD minimax modeling file.

Change 2: small nits: let's call this `extra_kwargs`. So that later on any other `gfx` arch, we can pass in their own arguments as well.

**Comment by hongxiayang** [comment]
verified it does help perf on gfx942. 
@yueliu14 : i tested on gfx950, and posted a pr to extend this PR: https://github.com/vllm-project/vllm/pull/46546. 
We can either land your PR first, then mine, or consolidate.

File: vllm/models/minimax_m3/common/ops/sparse_attn.py (hunk 2 above)
Comment (yueliu14) [reply vllm/models/minimax_m3/common/ops/sparse_attn.py]: hi @tjtanaa , thanks for reviewing! I have moved files to amd/ops, added import router and re-tested. Please take a look

**Review by hongxiayang — APPROVED** [review]

File: vllm/models/minimax_m3/common/ops/sparse_attn.py (hunk 2 above)
Comment (dllehr-amd) [reply vllm/models/minimax_m3/common/ops/sparse_attn.py]: @yueliu14  can you also change the gfx942_kwargs, to the extra_kwargs?

File: vllm/models/minimax_m3/amd/ops/index_topk.py [hunk 3]
```diff
@@ -0,0 +1,928 @@
...
+
+@torch.no_grad()
+def minimax_m3_index_topk(
+    score: torch.Tensor,  # [num_idx_heads, total_q, max_block]
+    cu_seqlens_q: torch.Tensor,  # [batch+1] int32
+    prefix_lens: torch.Tensor,  # [batch] int32
+    max_query_len: int,
+    topk: int,
+    init_blocks: int,
+    local_blocks: int,
```
Comment (dllehr-amd) [inline vllm/models/minimax_m3/amd/ops/index_topk.py:725]: why did you remove the `out: torch.Tensor | None = None,` code from here?  Wouldn't this cause an argument error as soon as we tried to run it?

**Review by dllehr-amd — CHANGES_REQUESTED** [review]
Please confirm the keyword error in `minimax_m3_index_topk` is gone, but also a bigger nit.  You completely copied all of the common code, and then added the gfx942 check.  So any changes to the common code won't benefit anyone other than gfx942.  Can you please only break out the gfx942 code into amd/ops for now?  and leave what can be common as common?
